src/ccl_musigma.c

```diff
@@ -184,85 +128,187 @@ void ccl_cosmology_spline_linpower_musigma(ccl_cosmology* cosmo, ccl_f2d_t *psp,
+            if(spstatus) {
+              *status = CCL_ERROR_MEMORY;
+              ccl_cosmology_set_status_message(cosmo,
+                                               "ccl_musigma.c: ccl_rescale_musigma_s8(): "
+                                               "Error initializing spline\n");
+            }
+          }
+          if(*status==0) {
+            gsl_spline_free(psp->fa);
+            psp->fa=fa;
+          }
+        }
+        else {
+          gsl_spline2d *fka = gsl_spline2d_alloc(gsl_interp2d_bicubic,
+                                                 psp->fka->interp_object.xsize,
+                                                 psp->fka->interp_object.ysize); 
+          if(fka == NULL) {
+            *status = CCL_ERROR_MEMORY;
+            ccl_cosmology_set_status_message(cosmo,
+                                             "ccl_musigma.c: ccl_rescale_musigma_s8(): "
+                                             "memory allocation\n");
+          }
+          if(*status==0) {
+            int spstatus = gsl_spline2d_init(fka, psp->fka->xarr,
+                                             psp->fka->yarr, psp->fka->zarr,
+                                             psp->fka->interp_object.xsize,
+                                             psp->fka->interp_object.ysize);
+            if(spstatus) {
+              *status = CCL_ERROR_MEMORY;
+              ccl_cosmology_set_status_message(cosmo,
+                                               "ccl_musigma.c: ccl_rescale_musigma_s8(): "
+                                               "Error initializing spline\n");
+            }
+          }
+          if(*status==0) {
+            gsl_spline2d_free(psp->fka);
+            psp->fka=fka;
+          }
         }
       }
+      free(D_mu);
+      free(D_GR);
+      ccl_parameters_free(&params_GR);
+      ccl_cosmology_free(cosmo_GR);
     }
   }
 
+  // if desried, renomalize to a given sigma8
   if (*status == 0) {
-    cosmo->data.p_lin = ccl_f2d_t_new(
-      na, aa, nk, lk, lpk_ln, NULL, NULL, 0,
-      1, 2, ccl_f2d_cclgrowth, 1, NULL, 0, 2,
-      ccl_f2d_3,status);
-  }
+    if (isfinite(cosmo->params.sigma8) && (!isfinite(cosmo->params.A_s))) {
+      size_t na, nk;
+      double sigma8, renorm=1;
 
-  // if desried, renomalize to a given sigma8
-  if (isfinite(cosmo->params.sigma8) && (!isfinite(cosmo->params.A_s))) {
-    if (*status == 0) {
-      cosmo->computed_linear_power = true;
-      sigma8 = ccl_sigma8(cosmo, status);
-      cosmo->computed_linear_power = false;
-    }
+      sigma8 = ccl_sigma8(cosmo, psp, status);
 
-    if (*status == 0) {
-      // Calculate normalization factor using computed value of sigma8, then
-      // recompute P(k, a) using this normalization
-      log_sigma8 = 2*(log(cosmo->params.sigma8) - log(sigma8));
-      for(int j = 0; j<na*nk; j++)
-        lpk_ln[j] += log_sigma8;
-    }
+      if (*status == 0) {
```

Comment:
There is a lot of somewhat duplicated spline resetting code here. It might be good to add an extra function in the spline class that does all of this there to avoid duplication and make long-term changes to the spline stuff easier.

Reply:
I've streamlined this function so the resplining only happens once. The previous version was still relying too much on the old code. May be worth looking at the file again rather than looking at the diff... :-(

Reply:
looks good!